### PR TITLE
Issue #89 Policy based access control for SAML 2.0 IdP (new feature)

### PR DESCRIPTION
**Pull request by Iwakata**
## Analysis
See the description section of the Issue #89.
## Solution

See the solution mentioned in the Issue #89.

This feature is made up of three parts described below:

1. An implementation of `SAML2IdentityProviderAdapter` interface.  Its `preSendResponse` method checks user/session properties against defined policies.

2. A definition of service configuration called `AuthProviderService`.  It is used for creating access policies .

3. Additional parameters used for policy evaluation.  Some env conditions, namely `LDAP filter condition` and `Membership condition` require realm or subject parameters.

## Install/Update

A simple configuration guide is attached with this request.

## Compatibility

None.

## Performance

Performance might be affected if complicated access policies need to be applied.

## I18N

None.

## Testing

- Set up an IdP and an SP.
- Set up the adapter and define access policies (resource, subject conditions, env conditions).
- Request an assertion from the SP.

Depending on the user/session properties, the request is:

- denied or
- accepted and an assertion is issued or
- processed further and session upgrade is performed.

## Regression testing

None.

### CODE REVIEW

**Comment by Iwakata** [comment]
Configuration Guide (in Japanese) is attached here.
[ConfigSAML2Policy.pdf](https://github.com/openam-jp/openam/files/3689471/ConfigSAML2Policy.pdf)

File: openam-federation/openam-federation-library/src/main/java/jp/co/osstech/oam/saml2/plugins/PolicyCheckIDPAdapter.java [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+import com.sun.identity.saml2.protocol.AuthnRequest;
+import com.sun.identity.saml2.protocol.Response;
+import com.sun.identity.security.AdminTokenAction;
+import com.sun.identity.shared.Constants;
+import com.sun.identity.shared.configuration.SystemPropertiesManager;
+import com.sun.identity.shared.encode.URLEncDec;
+import com.sun.identity.sm.DNMapper;
+import java.io.IOException;
+import java.security.AccessController;
+import java.util.Collections;
```
Comment (tsujiguchitky) [inline openam-federation/openam-federation-library/src/main/java/jp/co/osstech/oam/saml2/plugins/PolicyCheckIDPAdapter.java]: `java.util.Collections` is not used.

File: openam-federation/openam-federation-library/src/main/java/jp/co/osstech/oam/saml2/plugins/PolicyCheckIDPAdapter.java [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+
+import com.iplanet.sso.SSOToken;
+import com.sun.identity.authentication.util.AMAuthUtils;
+import com.sun.identity.authentication.util.ISAuthConstants;
+import com.sun.identity.entitlement.Entitlement;
+import com.sun.identity.entitlement.EntitlementException;
+import com.sun.identity.entitlement.Evaluator;
+import com.sun.identity.entitlement.opensso.SubjectUtils;
+import com.sun.identity.federation.common.FSUtils;
+import com.sun.identity.policy.interfaces.Condition;
```
Comment (tsujiguchitky) [inline openam-federation/openam-federation-library/src/main/java/jp/co/osstech/oam/saml2/plugins/PolicyCheckIDPAdapter.java]: `com.sun.identity.policy.interfaces.Condition` is a deprecated class. How about replacing with `org.forgerock.openam.entitlement.conditions.environment.ConditionConstants` ?

Comment (Iwakata) [reply openam-federation/openam-federation-library/src/main/java/jp/co/osstech/oam/saml2/plugins/PolicyCheckIDPAdapter.java]: I agree.

File: openam-federation/openam-federation-library/src/main/java/jp/co/osstech/oam/saml2/plugins/PolicyCheckIDPAdapter.java (hunk 1 above)
Comment (Iwakata) [reply openam-federation/openam-federation-library/src/main/java/jp/co/osstech/oam/saml2/plugins/PolicyCheckIDPAdapter.java]: Understood.

**Pull request merged**